Userland/Libraries/LibGfx/TextLayout.cpp

```diff
@@ -7,51 +7,85 @@
+static DrawGlyphOrEmoji construct_glyph_or_emoji(size_t index, FloatPoint const& position, Gfx::Font const& font, Span<hb_glyph_info_t const> glyph_info, Span<hb_glyph_info_t const> input_glyph_info)
 {
-    u32 code_point = *it;
-    auto next_code_point = it.peek(1);
+    if (font.has_color_bitmaps()) {
+        auto cluster_start = glyph_info[index].cluster;
+        auto cluster_end = [&]() -> u32 {
+            if (index + 1 < glyph_info.size())
+                return glyph_info[index + 1].cluster;
+            return input_glyph_info.last().cluster + 1;
+        }();
 
-    ScopeGuard consume_variation_selector = [&, initial_it = it] {
-        // If we advanced the iterator to consume an emoji sequence, don't look for another variation selector.
-        if (initial_it != it)
-            return;
+        Vector<u32> cluster;
+        for (size_t j = 0; j < input_glyph_info.size(); ++j) {
+            auto const& glyph = input_glyph_info[j];
+            if (glyph.cluster >= cluster_end)
+                break;
+            if (glyph.cluster >= cluster_start)
+                cluster.append(glyph.codepoint);
+        }
 
-        // Otherwise, discard one code point if it's a variation selector.
-        if (next_code_point.has_value() && Unicode::code_point_has_variation_selector_property(*next_code_point))
-            ++it;
+        if (auto const* emoji = Emoji::emoji_for_code_points(cluster)) {
+            return DrawEmoji {
+                .position = position,
+                .emoji = emoji,
+            };
+        }
+    }
+
+    return DrawGlyph {
+        .position = position,
+        .glyph_id = glyph_info[index].codepoint,
     };
+}
 
-    // NOTE: We don't check for emoji
-    auto font_contains_glyph = font.contains_glyph(code_point);
-    auto check_for_emoji = !font.has_color_bitmaps() && Unicode::could_be_start_of_emoji_sequence(it, font_contains_glyph ? Unicode::SequenceType::EmojiPresentation : Unicode::SequenceType::Any);
+void for_each_glyph_position(FloatPoint baseline_start, Utf8View string, Gfx::Font const& font, Function<void(DrawGlyphOrEmoji const&)> callback, IncludeLeftBearing include_left_bearing, Optional<float&> width)
+{
+    hb_buffer_t* buffer = hb_buffer_create();
+    hb_buffer_add_utf8(buffer, reinterpret_cast<char const*>(string.bytes()), string.byte_length(), 0, -1);
+    hb_buffer_guess_segment_properties(buffer);
 
-    // If the font contains the glyph, and we know it's not the start of an emoji, draw a text glyph.
-    if (font_contains_glyph && !check_for_emoji) {
-        return DrawGlyph {
-            .position = point,
-            .code_point = code_point,
-        };
+    u32 glyph_count;
+    auto* glyph_info = hb_buffer_get_glyph_infos(buffer, &glyph_count);
+    Vector<hb_glyph_info_t> const input_glyph_info({ glyph_info, glyph_count });
+    if (input_glyph_info.is_empty()) {
+        hb_buffer_destroy(buffer);
```

Comment:
nit: using AK::ScopeGuard at the top of a scope might a better idea.

Reply:
It's using a scope guard now.